libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm

```diff
@@ -0,0 +1,214 @@
+;*****************************************************************************
+;* x86-optimized functions for dnn native backend convolution
+;*
+;* This file is part of FFmpeg.
+;*
+;* FFmpeg is free software; you can redistribute it and/or
+;* modify it under the terms of the GNU Lesser General Public
+;* License as published by the Free Software Foundation; either
+;* version 2.1 of the License, or (at your option) any later version.
+;*
+;* FFmpeg is distributed in the hope that it will be useful,
+;* but WITHOUT ANY WARRANTY; without even the implied warranty of
+;* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+;* Lesser General Public License for more details.
+;*
+;* You should have received a copy of the GNU Lesser General Public
+;* License along with FFmpeg; if not, write to the Free Software
+;* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+;******************************************************************************
+
+%include "libavutil/x86/x86util.asm"
+
+SECTION .text
+
+%macro COUNT_INPUT 0
+    mov tmp1d, padding_method
+    cmp tmp1d, SAME_CLAMP_TO_EDGE
+    je .clamp
+
+    cmp y_posd, 0
+    jl .out_of_th
+    mov tmp2d, height
+    cmp y_posd, tmp2d
+    jge .out_of_th
+
+    cmp x_posd, 0
+    jl .out_of_th
+    mov tmp2d, width
+    cmp x_posd, tmp2d
+    jge .out_of_th
+
+    mov tmp1d, y_posd
+    imul tmp1d, src_linesize
+    mov tmp2d, x_posd
+    imul tmp2d, input_num
+    add tmp1d, tmp2d
+    jmp .count_end
+
+    .out_of_th:
+        mov tmp1d, -1
+        jmp .count_end
+
+    .clamp:
+    cmp y_posd, 0
+    jl .y_clamp_zero
+    mov tmp1d, height
+    cmp y_posd, tmp1d
+    jge .y_clamp_height
+    mov tmp1d, y_posd
+    jmp .y_normal
+
+    .y_clamp_zero:
+        xor tmp1d, tmp1d
+        jmp .y_normal
+
+    .y_clamp_height:
+        sub tmp1d, 1
+
+    .y_normal:
+
+    cmp x_posd, 0
+    jl .x_clamp_zero
+    mov tmp2d, width
+    cmp x_posd, tmp2d
+    jge .x_clamp_width
+    mov tmp2d, x_posd
+    jmp .x_normal
+
+    .x_clamp_zero:
+        xor tmp2d, tmp2d
+        jmp .x_normal
+
+    .x_clamp_width:
+        sub tmp2d, 1
+
+    .x_normal:
+
+    imul tmp1d, src_linesize
+    imul tmp2d, input_num
+    add tmp1d, tmp2d
+
+    .count_end:
+%endmacro
+
+; void ff_dnn_execute_layer_conv2d_sse4(const execute_param *exe_param);
+
+%if ARCH_X86_64
+INIT_XMM sse4
+cglobal dnn_execute_layer_conv2d, 8, 15, 3, exe_param,\
+    x, y, n_filter, cha, kernel_x, kernel_y, x_pos, y_pos, kernel_pos,\
+    input, output, kernel, tmp1, tmp2
```

Comment:
%macro cglobal 1-2+ "" ; name, [PROLOGUE args]

why we need to add x, y, n_filter, ... which are not function args, they are also not used in the below code.

Reply:
x, y, n_filter were names of registers, which can be seen as variables.

Reply:
but there variables are not used in the below code.